File: Assignment 2/Python/1_question.py
"""
1. Write a program to sort a list of element in an Array using Quick Sort. Show
the running time complexity w.r.t different input cases (best/average/worst).
"""

from timeit import default_timer as timer


def partition(array, low, high):
    pivot = array[high]
    i = low - 1
    for j in range(low, high):
        if array[j] <= pivot:
            i = i + 1
            (array[i], array[j]) = (array[j], array[i])
    (array[i + 1], array[high]) = (array[high], array[i + 1])
    return i + 1


def quick_sort(array, low, high):
    if low < high:
        pi = partition(array, low, high)
        quick_sort(array, low, pi - 1)
        quick_sort(array, pi + 1, high)


data = [1, 7, 4, 1, 10, 9, -2]
print("Unsorted Array")
print(data)

size = len(data)

start = timer()
quick_sort(data, 0, size - 1)
end = timer()

print('Sorted Array in Ascending Order:')
print(data)
print(end - start)

File: Assignment 2/C/1_question.c
/*
1. Write a program to sort a list of element in an Array using Quick Sort. Show
the running time complexity w.r.t different input cases (best/average/worst).
*/

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void swap(int* a, int* b)
{
    int t = *a;
    *a = *b;
    *b = t;
}

int partition(int arr[], int low, int high)
{
    int pivot = arr[high];
    int i = low - 1;
 
    for (int j = low; j <= high - 1; ++j)
    {
        if (arr[j] < pivot)
        {
            ++i;
            swap(&arr[i], &arr[j]);
        }
    }

    swap(&arr[i + 1], &arr[high]);
    return (i + 1);
}

void quickSort(int arr[], int low, int high)
{
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

void printArray(int arr[], int size)
{
    int i;
    for (i = 0; i < size; ++i)
        printf("%d ", arr[i]);
    
    printf("\n");
}

int main()
{
    int arr[] = { 10, 7, 8, 9, 1, 5 };
    clock_t start, end;
    double time_taken;
    
    int n = sizeof(arr) / sizeof(arr[0]);

    start = clock();
    quickSort(arr, 0, n - 1);
    end = clock();

    printf("The sorted array: ");
    printArray(arr, n);

    time_taken = (double)(end - start) / CLOCKS_PER_SEC;
    printf("\n The elapsed time is %f seconds \n", time_taken);

    return 0;
}